docs/tuf-artifact-replication.adoc

```diff
@@ -0,0 +1,185 @@
+dataset on each M.2 device: the file name of each stored artifact is its
+SHA-256 hash.
+
+It also stores an _artifact configuration_ in memory: a list of all
+artifact hashes that the sled should store, and a generation number.
+The generation number is owned by Nexus, which increments the generation
+number when the set of TUF repos on the system changes. Sled Agent
+prevents modifying the configuration without an increase in the
+generation number.
+
+Sled Agent offers the following APIs on the underlay network, intended
+for Nexus:
+
+* `artifact_config_get`: Get the current artifact configuration.
+* `artifact_config_put`: Put the artifact configuration that should be
+  in effect. This API is idempotent (putting the same configuration does
+  not change anything). Modified configurations must also increase the
+  generation number.
+* `artifact_list`: List the artifacts present in the artifact
+  configuration along with the count of available copies of each
+  artifact across the *update* datasets. Also includes the current
+  generation number.
+* `artifact_put`: Put the request body into the artifact store.
+  Rejects the request if the artifact does not belong to the current
+  configuration.
+* `artifact_copy_from_depot`: Sends a request to another Sled Agent (via
+  the *TUF Repo Depot API*; see below) to fetch an artifact. The base
+  URL for the source sled is chosen by the requester. This API responds
+  after a successful HTTP response from the source sled and the copy
+  proceeds asynchronously. Rejects the request if the artifact does not
+  belong to the current configuration.
+
+Sled Agent also spawns another Dropshot API server called the *TUF Repo
+Depot API* which offers one API on the underlay network, intended for
+other Sled Agents:
+
+* `artifact_get_by_sha256`: Get the content of an artifact.
+
+In an asynchronous task called the _delete reconciler_, Sled Agent
+periodically scans the *update* datasets for artifacts that are not
+part of the present configuration and deletes them. Prior to each
+filesystem operation the task checks the configuration for presence of
+that artifact hash. The delete reconciler then waits for an artifact
+configuration change until running again.
+
+== Nexus implementation
+
+Nexus has a `tuf_artifact_replication` background task which runs this
+reliable persistent workflow:
+
+1. Collect the artifact configuration (the list of artifact hashes, and
+   the current generation number) from the database.
+2. Call `artifact_config_put` on all sleds. Stop if any sled rejects the
+   configuration (our information is already out of date).
+3. Call `artifact_list` on all sleds. Stop if any sled informs us of a
+   different generation number.
+4. Delete any local copies of repositories where all artifacts are
+   sufficiently replicated across sleds. ("Sufficiently replicated"
+   currently means that at least 3 sleds each have at least one copy.)
```

Comment:
When would sled have more than one copy of an artifact?

Reply:
Artifacts are stored on both M.2 devices.

Reply:
Aha. That makes sense. Thanks.